src/result_enumerator.rs

```diff
@@ -2,99 +2,73 @@ use crate::{
-    pub inner: NonNull<IWbemClassObject>,
+    pub inner: IWbemClassObject,
 }
 
 impl IWbemClassWrapper {
-    pub unsafe fn new(ptr: NonNull<IWbemClassObject>) -> Self {
-        Self { inner: ptr, }
-    }
-
-    /// Creates a copy of the pointer and calls
-    /// [AddRef](https://docs.microsoft.com/en-us/windows/win32/api/unknwn/nf-unknwn-iunknown-addref)
-    /// to increment Reference Count.
-    ///
-    /// # Safety
-    /// See [Managing the lifetime of an object](https://docs.microsoft.com/en-us/windows/win32/learnwin32/managing-the-lifetime-of-an-object)
-    /// and [Rules for managing Ref count](https://docs.microsoft.com/en-us/windows/win32/com/rules-for-managing-reference-counts)
-    ///
-    pub unsafe fn clone(ptr: NonNull<IWbemClassObject>) -> Self {
-        let refcount = ptr.as_ref().AddRef();
-        trace!("Reference count: {}", refcount);
-        Self::new(ptr)
+    pub fn new(inner: IWbemClassObject) -> Self {
+        Self { inner }
     }
 
     /// Return the names of all the properties of the given object.
     ///
     pub fn list_properties(&self) -> WMIResult<Vec<String>> {
-        // This will store the properties names from the GetNames call.
-        let mut p_names = NULL as *mut SAFEARRAY;
-
-        let ptr = self.inner.as_ptr();
-
-        unsafe {
-            check_hres((*ptr).GetNames(
-                ptr::null(),
-                (WBEM_FLAG_ALWAYS | WBEM_FLAG_NONSYSTEM_ONLY) as i32,
+        let p_names = unsafe {
+            self.inner.GetNames(
+                None,
+                WBEM_FLAG_ALWAYS.0 | WBEM_FLAG_NONSYSTEM_ONLY.0,
                 ptr::null_mut(),
-                &mut p_names,
-            ))?;
+            )
+        }?;
 
-            let res = safe_array_to_vec_of_strings(p_names);
+        let res = safe_array_to_vec_of_strings(unsafe { &*p_names });
 
-            check_hres(SafeArrayDestroy(p_names))?;
+        unsafe { SafeArrayDestroy(p_names) }?;
 
-            res
-        }
+        res
     }
 
     pub fn get_property(&self, property_name: &str) -> WMIResult<Variant> {
-        let name_prop = BStr::from_str(property_name)?;
+        let name_prop = HSTRING::from(property_name);
 
-        let mut vt_prop: VARIANT = unsafe { mem::zeroed() };
+        let mut vt_prop = VARIANT::default();
 
         let mut cim_type = 0;
 
         unsafe {
-            check_hres((*self.inner.as_ptr()).Get(
-                name_prop.as_lpcwstr(),
+            self.inner.Get(
+                &name_prop,
                 0,
                 &mut vt_prop,
                 &mut cim_type,
                 ptr::null_mut(),
-            ))?;
+            )?;
 
             let property_value =
-                Variant::from_variant(vt_prop)?.convert_into_cim_type(cim_type as _)?;
+                Variant::from_variant(&vt_prop)?.convert_into_cim_type(CIMTYPE_ENUMERATION(cim_type))?;
 
-            check_hres(VariantClear(&mut vt_prop))?;
+            VariantClear(&mut vt_prop)?;
```

Comment:
~Do we need this now? Maybe `VARIANT` does this by itself?~
No, there's no `Drop` impl

Reply:
Yes, windows-rs does not improve the ease of use for VARIANT ^^ It's quite a pain to deal with, because it can be of any type and filled in many different ways. there's no clone/drop or any helpers for it yet in this crate.